### PR TITLE
LPS-184642, LPS-185868

### PR DESCRIPTION
**Pull request by ces-totrinh**
Hi @locpham97, @ces-quanhuynh, @ces-luannguyen,
Please help review this PR and leave your comments.

Motivation
==========
[LPS-184642](https://liferay.atlassian.net/browse/LPS-184642) [FE] Add a checkbox to set/unset the propagation of permissions on the folder permission modal

Solution
==========
The Permissions method is in use in many places not only for DLFolder, so we are looking for a general solution without hard-coding. And the idea is:
We can create an interface class and a Base class that implements the interface. Then we could create a service component for DLFolder that extends the Base class and could be targeted by portlet name.
After that, modify the edit_permission file and use the service to add a checkbox.
btw, the checkbox is shown only on the Folder Permission modal. not shown on DLFileEntry and Bulk permission.
Thanks.

### CODE REVIEW

**Review by ces-quanhuynh — APPROVED** [review]

File: modules/apps/document-library/document-library-web/src/main/java/com/liferay/document/library/web/internal/portlet/configuration/permission/propagation/DLFolderPortletConfigurationPermissionPropagation.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+	@Override
+	public void updatePermissionPropagation(
+		ActionRequest actionRequest, ActionResponse actionResponse) {
+	}
+
+	private boolean _isBulkSelection(PortletRequest portletRequest) {
+		String[] resourcePrimKeys = ParamUtil.getStringValues(
+			portletRequest, "resourcePrimKey");
+
+		if (resourcePrimKeys.length > 1) {
```
Comment (ces-quanhuynh) [inline modules/apps/document-library/document-library-web/src/main/java/com/liferay/document/library/web/internal/portlet/configuration/permission/propagation/DLFolderPortletConfigurationPermissionPropagation.java:111]: We should return here too.

File: modules/apps/portlet-configuration/portlet-configuration-web/src/main/resources/META-INF/resources/edit_permissions.jsp [hunk 2]
```diff
@@ -227,11 +229,24 @@ if (Validator.isNotNull(portletConfigurationPermissionsDisplayContext.getModelRe
 		</aui:form>
 	</div>
 
-	<aui:button-row>
-		<aui:button name="saveButton" type="submit" />
-
-		<aui:button type="cancel" />
-	</aui:button-row>
+	<div class="bg-white fixed-bottom modal-footer">
+		<c:if test="<%= (portletConfigurationPermissionPropagation != null) && portletConfigurationPermissionPropagation.isShow(renderRequest) %>">
```
Comment (ces-quanhuynh) [inline modules/apps/portlet-configuration/portlet-configuration-web/src/main/resources/META-INF/resources/edit_permissions.jsp:233]: I think we should use Validator.isNotNull

File: modules/apps/document-library/document-library-web/src/main/java/com/liferay/document/library/web/internal/portlet/configuration/permission/propagation/DLFolderPortletConfigurationPermissionPropagation.java [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+			classPK = DLFolderConstants.DEFAULT_PARENT_FOLDER_ID;
+		}
+
+		PermissionPropagationEntry permissionPropagationEntry =
+			_permissionPropagationEntryLocalService.
+				fetchPermissionPropagationEntry(
+					themeDisplay.getCompanyId(), groupId,
+					DLFolder.class.getName(), classPK);
+
+		if (permissionPropagationEntry != null) {
```
[the remark this answers is not in the document]
Comment (tototrinh) [reply modules/apps/document-library/document-library-web/src/main/java/com/liferay/document/library/web/internal/portlet/configuration/permission/propagation/DLFolderPortletConfigurationPermissionPropagation.java:77]: I don't think we need to use `Validator.isNotNull()` for permissionPropagationEntry. See [here](https://github.com/liferay/liferay-portal/blob/master/portal-kernel/src/com/liferay/portal/kernel/util/Validator.java#L912-L913)

File: modules/apps/portlet-configuration/portlet-configuration-web/src/main/resources/META-INF/resources/edit_permissions.jsp (hunk 2 above)
Comment (tototrinh) [reply modules/apps/portlet-configuration/portlet-configuration-web/src/main/resources/META-INF/resources/edit_permissions.jsp:233]: Same above. No need use Validator.isNotNull

File: modules/apps/document-library/document-library-web/src/main/java/com/liferay/document/library/web/internal/portlet/configuration/permission/propagation/DLFolderPortletConfigurationPermissionPropagation.java (hunk 1 above)
Comment (tototrinh) [reply modules/apps/document-library/document-library-web/src/main/java/com/liferay/document/library/web/internal/portlet/configuration/permission/propagation/DLFolderPortletConfigurationPermissionPropagation.java:111]: This follows the Liferay convention. Thanks.

File: modules/apps/document-library/document-library-web/src/main/java/com/liferay/document/library/web/internal/portlet/configuration/permission/propagation/DLFolderPortletConfigurationPermissionPropagation.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+		if (classPK == DLFolderConstants.DEFAULT_PARENT_FOLDER_ID) {
+			return true;
+		}
+
+		return false;
+	}
+
+	@Override
+	public boolean isShow(PortletRequest portletRequest) {
+		if ((!Objects.equals(
```
[the remark this answers is not in the document]
Comment (tototrinh) [reply modules/apps/document-library/document-library-web/src/main/java/com/liferay/document/library/web/internal/portlet/configuration/permission/propagation/DLFolderPortletConfigurationPermissionPropagation.java:90]: This follows the Liferay convention. Thanks.

File: portal-impl/src/com/liferay/portal/service.xml [hunk 5]
```diff
@@ -1442,6 +1442,12 @@
 		<column name="classNameId" type="long" />
 		<column name="classPK" type="long" />
 		<column name="propagation" type="boolean" />
+		<finder name="G_C_C_C" return-type="PermissionPropagationEntry" unique="true">
```
[the remark this answers is not in the document]
Comment (tototrinh) [reply portal-impl/src/com/liferay/portal/service.xml:1448]: Added. Thanks.

**Review by ces-locpham — APPROVED** [review]

**Review by ces-quanhuynh — APPROVED** [review]

**Comment by ces-totrinh** [comment]
@hhtank, this PR passed our internal review. Please help take a look. Thanks.

**Comment by hhtank** [comment]
@tototrinh 
LGTM, thanks!

**Comment by ces-totrinh** [comment]
I update the PR to fix the conflict. Thanks.

**Comment by ces-totrinh** [comment]
Merged. Thanks all.

**Pull request merged**